packages/app-store/larkcalendar/api/callback.ts

```diff
@@ -0,0 +1,95 @@
+import type { NextApiRequest, NextApiResponse } from "next";
+
+import { getSafeRedirectUrl } from "@calcom/lib/getSafeRedirectUrl";
+import logger from "@calcom/lib/logger";
+import prisma from "@calcom/prisma";
+
+import { decodeOAuthState } from "../../_utils/decodeOAuthState";
+import { LARK_HOST } from "../common";
+import { getAppAccessToken } from "../lib/AppAccessToken";
+import type { LarkAuthCredentials } from "../types/LarkCalendar";
+
+const log = logger.getChildLogger({ prefix: [`[[lark/api/callback]`] });
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  const { code } = req.query;
+
+  if (typeof code !== "string") {
+    res.status(400).json({ message: "No code returned" });
+    return;
+  }
+
+  const state = decodeOAuthState(req);
+
+  try {
+    const appAccessToken = await getAppAccessToken();
+
+    const response = await fetch(`https://${LARK_HOST}/open-apis/authen/v1/access_token`, {
+      method: "POST",
+      headers: {
+        Authorization: "Bearer " + appAccessToken,
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify({
+        grant_type: "authorization_code",
+        code,
+      }),
+    });
+
+    const responseBody = await response.json();
+
+    if (!response.ok || responseBody.code !== 0) {
+      log.error("get user_access_token failed with none 0 code", responseBody);
+      return res.redirect("/apps/installed?error=" + JSON.stringify(responseBody));
+    }
+
+    const key: LarkAuthCredentials = {
+      expiry_date: Math.round(+new Date() / 1000 + responseBody.data.expires_in),
+      access_token: responseBody.data.access_token,
+      refresh_token: responseBody.data.refresh_token,
+      refresh_expires_date: Math.round(+new Date() / 1000 + responseBody.data.refresh_expires_in),
+    };
+
+    /**
+     * A user can have only one pair of refresh_token and access_token effective
+     * at same time. Newly created  refresh_token and access_token will invalidate
+     * older ones. So we need to keep only one lark credential per user only.
+     * A user may connect many times, since both userId and type are
+     * not unique in schema, so we cannot use upsert
+     */
+    const currentCredential = await prisma.credential.findFirst({
+      where: {
+        userId: req.session?.user.id,
+        type: "lark_calendar",
+      },
+    });
+
+    if (!currentCredential) {
+      await prisma.credential.create({
+        data: {
+          type: "lark_calendar",
+          key,
+          userId: req.session?.user.id,
+          appId: "lark-calendar",
+        },
+      });
+    } else {
+      await prisma.credential.update({
+        data: {
+          type: "lark_calendar",
+          key,
+          userId: req.session?.user.id,
+          appId: "lark-calendar",
+        },
+        where: {
+          id: currentCredential.id,
+        },
+      });
+    }
```

Comment:
Yes of cause

Reply:
I'm afraid upsert cannot work here. 

A user can have only one pair of refresh_token and access_token effective at same time. Newly created  refresh_token and access_token will invalidate older ones. So we need to keep only one lark credential per user only.

However, a user may connect many times, since both userId and type are not unique in schema, so we have to use credential id as index for looking for the unique access_token token.

in this case, id does not exist before created, so we cannot use credential id (which may not exist) as where statement

Reply:
larkcalendar/api/events.ts is for

1. handle event subscription, see https://open.larksuite.com/document/ukTMukTMukTM/uUTNz4SN1MjL1UzM

2. handle  app ticket (which is used for retrieve access token for calling event related api) event callback from lark. see https://open.larksuite.com/document/uAjLw4CM/ukTMukTMukTM/application-v6/event/app_ticket-events and https://open.larksuite.com/document/ukTMukTMukTM/ukDNz4SO0MjL5QzM/auth-v3/auth/app_access_token
handle bot related events.

purpose of larkcalendar/lib/BotService/ts 
1. If calcom is registered as webapp in lark. calcom have to able to login with lark account

2. to avoid 1, we change our integration type as bot. (A bot will be shown in lark)

3. every time a user   communicates with the bot, bot will always send message to user